aws/network-primitives/instances.tf

```diff
@@ -0,0 +1,297 @@
+resource "aws_instance" "bootstrap_node" {
+  count             = length(var.aws_region) * var.instance_count
+  ami               = data.aws_ami.ubuntu_amd64.image_id
+  instance_type     = var.instance_type
+  subnet_id         = element(aws_subnet.public_subnets.*.id, count.index)
+  availability_zone = element(var.azs, count.index)
+  # Security Group
+  vpc_security_group_ids = ["${aws_security_group.gemini_sg.id}"]
+  # the Public SSH key
+  key_name                    = var.aws_key_name
+  associate_public_ip_address = true
+  ebs_optimized = true
+  ebs_block_device {
+    device_name = "/dev/sda1"
+    volume_size = 100
+  }
+
+
+  tags = {
+    name       = "${var.network_name}-bootstrap"
+    role       = "bootstrap node"
+    os_name    = "ubuntu"
+    os_version = "22.04"
+    arch       = "x86_64"
+  }
+
+  depends_on = [
+    aws_subnet.public_subnets,
+    aws_nat_gateway.nat_gateway,
+    aws_internet_gateway.gw
+  ]
+
+  lifecycle {
+
+    create_before_destroy = true
+
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "sleep 180",
+      "sudo apt update -y",
+      "sudo DEBIAN_FRONTEND=noninteractive apt install curl gnupg openssl net-tools -y",
+    ]
+
+    on_failure = continue
+
+  }
+
+  # Setting up the ssh connection
+  connection {
+    type        = "ssh"
+    host        = element(self.*.public_ip, count.index)
+    user        = "ubuntu"
+    private_key = file("${var.private_key_path}")
+    timeout     = "90s"
+  }
+
+}
+
+
+resource "aws_instance" "full_node" {
+  count             = length(var.aws_region) * var.instance_count
+  ami               = data.aws_ami.ubuntu_amd64.image_id
+  instance_type     = var.instance_type
+  subnet_id         = element(aws_subnet.public_subnets.*.id, count.index)
+  availability_zone = element(var.azs, count.index)
+  # Security Group
+  vpc_security_group_ids = ["${aws_security_group.gemini_sg.id}"]
+  # the Public SSH key
+  key_name                    = var.aws_key_name
+  associate_public_ip_address = true
+  ebs_optimized = true
+  ebs_block_device {
+    device_name = "/dev/sda1"
+    volume_size = 100
+  }
+
+  tags = {
+    name       = "${var.network_name}-full"
+    role       = "full node"
+    os_name    = "ubuntu"
+    os_version = "22.04"
+    arch       = "x86_64"
+  }
+
+  depends_on = [
+    aws_subnet.public_subnets,
+    aws_nat_gateway.nat_gateway,
+    aws_internet_gateway.gw
+  ]
+
+  lifecycle {
+
+    create_before_destroy = true
+
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "sleep 180",
```

Comment:
:thinking: why? This is the kind of thing that typically randomly fails in practice.

Reply:
It gives the cloud-init scripts time to finish before trying to run the remote provisioners else it will timeout, the aws launches go though an initialization stage before the health checks are done and they system state is "ready/running".

Reply:
Is there a way to detect when cloud-init finishes rather than having arbitrary timeouts?

180 seconds is A LOT, would be much better to proactively detect this rather than waiting for whole 3 minutes unnecessarily.

Reply:
There is no way for cloud-init to tell terraform it's finished. However, I have removed this and just increased the timeout for ssh connections.

Reply:
> There is no way for cloud-init to tell terraform it's finished. However, I have removed this and just increased the timeout for ssh connections.

It doesn't need to. `cloud-init` is doing something, right?  So you can have a loop here that will check for whatever you expect cloud-init to finish doing every few seconds and move forward only when it is done.

Removing delay and increasing timeout makes even less sense now. Will it automatically retry or something?